neuro_san/service/agent_server.py

```diff
@@ -106,48 +110,82 @@ def setup_tool_factory_provider(self):
+        service_router: DynamicAgentRouter = DynamicAgentRouter().get_instance()
+        service_router.remove_service(agent_service_name)
+
     def serve(self):
         """
         Start serving gRPC requests
         """
         values = agent_pb2.DESCRIPTOR.services_by_name.values()
-        server_lifetime = ServerLifetime(self.server_name,
-                                         self.server_name_for_logs,
-                                         self.port, self.logger,
-                                         request_limit=self.request_limit,
-                                         max_workers=self.max_concurrent_requests,
-                                         max_concurrent_rpcs=None,
-                                         # Used for health checking. Probably needs agent-specific love.
-                                         protocol_services_by_name_values=values,
-                                         loop_sleep_seconds=5.0,
-                                         server_loop_callbacks=self.server_loop_callbacks)
-
-        server = server_lifetime.create_server()
+        self.server_lifetime = ServerLifetime(
+            self.server_name,
+            self.server_name_for_logs,
+            self.port, self.logger,
+            request_limit=self.request_limit,
+            max_workers=self.max_concurrent_requests,
+            max_concurrent_rpcs=None,
+            # Used for health checking. Probably needs agent-specific love.
+            protocol_services_by_name_values=values,
+            loop_sleep_seconds=5.0,
+            server_loop_callbacks=self.server_loop_callbacks)
+
+        server = self.server_lifetime.create_server()
 
         # New-style service
-        security_cfg = None     # ... yet
+        self.security_cfg = None     # ... yet
 
-        self.setup_tool_factory_provider()
         tool_factory_provider: ServiceToolFactoryProvider = \
             ServiceToolFactoryProvider.get_instance()
+        # Add listener to handle adding per-agent gRPC services
+        # to our dynamic router:
+        tool_factory_provider.add_listener(self)
 
-        agent_names: List[str] = tool_factory_provider.get_agent_names()
-        for agent_name in agent_names:
-            service = AgentService(server_lifetime, security_cfg,
-                                   agent_name,
-                                   tool_factory_provider.get_agent_tool_factory_provider(agent_name),
-                                   self.server_logging)
-            self.services.append(service)
+        self.setup_tool_factory_provider()
 
-            servicer_to_server = AgentServicerToServer(service, agent_name=agent_name)
-            servicer_to_server.add_rpc_handlers(server)
+        # Add DynamicAgentRouter instance as a generic RPC handler for our server:
+        service_router: DynamicAgentRouter = DynamicAgentRouter().get_instance()
+        server.add_generic_rpc_handlers((service_router,))
 
```

Comment:
And DynamicAgentRouter instance is now our gRPC requests handler for AgentServer.